include/deal.II/base/tensor_base.h

```diff
@@ -359,70 +379,84 @@ class Tensor<1,dim,Number>
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
-   * Declare an array type which can be used to initialize statically an
-   * object of this type.
-   *
-   * Avoid warning about zero-sized array for <tt>dim==0</tt> by choosing
-   * lunatic value that is likely to overflow memory limits.
+   * The tensor type this object represents.
+   */
+  typedef Tensor<rank_,dim,Number> tensor_type;
+
+  /**
+   * Type of stored objects (i.e., the object returned by operator[]()).
+   * This is a tensor of lower rank.
+   */
+  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
```

Comment:
OK. It's a bit hard to see just from a diff, so if you are willing to clean this up so that we don't need the static casts any more, then I'm willing to let you off the hook for this patch. The issue with the static casts really bothers me enough that I think we should really make sure we don't need it.